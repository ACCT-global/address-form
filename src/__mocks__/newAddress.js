export default {
  addressId: '1',
  addressType: 'residential',
  city: null,
  complement: null,
  country: null,
  geoCoordinates: null,
  neighborhood: null,
  number: null,
  postalCode: null,
  receiverName: null,
  reference: null,
  state: null,
  street: null,
}
