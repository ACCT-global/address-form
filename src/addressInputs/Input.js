import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShapeWithValidation
  from '../propTypes/AddressShapeWithValidation'
import InputSelect from './InputSelect'
import InputText from './InputText'
import InputLabel from './InputLabel'
import PostalCodeLoader from '../postalCodeFrom/PostalCodeLoader'

class Input extends Component {
  render() {
    const { field, options, address, inputRef } = this.props
    const loading = !!address[field.name].loading

    if (field.name === 'postalCode') {
      return (
        <InputLabel field={field}>
          <InputText
            field={field}
            className={loading ? 'loading-postal-code' : null}
            address={address}
            onChange={this.props.onChange}
            onBlur={this.props.onBlur}
            inputRef={inputRef}
          />
          {loading && <PostalCodeLoader />}
        </InputLabel>
      )
    }

    return (
      <InputLabel field={field}>
        {options
          ? <InputSelect
            field={field}
            options={options}
            address={address}
            onChange={this.props.onChange}
            onBlur={this.props.onBlur}
            inputRef={inputRef}
            />
          : <InputText
            field={field}
            address={address}
            onChange={this.props.onChange}
            onBlur={this.props.onBlur}
            inputRef={inputRef}
            />}
      </InputLabel>
    )
  }
}

Input.defaultProps = {
  inputRef: () => {},
}

Input.propTypes = {
  field: PropTypes.object.isRequired,
  options: PropTypes.array,
  address: PropTypes.shape(AddressShapeWithValidation),
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  inputRef: PropTypes.func,
}

export default Input
