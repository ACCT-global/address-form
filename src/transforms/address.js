import reduce from 'lodash/reduce'
import filter from 'lodash/filter'
import map from 'lodash/map'
import difference from 'lodash/difference'
import find from 'lodash/find'
import findIndex from 'lodash/findIndex'
import { getField } from '../selectors/fields'
import { validateChangedFields } from '../validateAddress'
import msk from 'msk'

export function addValidation(address) {
  return reduce(
    address,
    (newAddress, value, propName) => {
      newAddress[propName] = { value }
      return newAddress
    },
    {}
  )
}

export function removeValidation(address) {
  return reduce(
    address,
    (newAddress, { value }, propName) => {
      newAddress[propName] = value
      return newAddress
    },
    {}
  )
}

export function addNewField(address, fieldName, value) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = {
        ...prop,
        [fieldName]: value,
      }
      return newAddress
    },
    {}
  )
}

export function removeField(address, fieldName) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = { ...prop }
      delete newAddress[propName][fieldName]
      return newAddress
    },
    {}
  )
}

export function addDisabledToProtectedFields(fields, rules) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasValue = prop && prop.value
      const isProtectField =
        rules.postalCodeProtectedFields &&
        rules.postalCodeProtectedFields.indexOf(propName) !== -1

      newFields[propName] = prop

      if (isProtectField && hasValue) {
        newFields[propName] = {
          ...prop,
          disabled: true,
        }
      }

      return newFields
    },
    {}
  )
}

export function handleMultipleValues(fields) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasMultipleValues = prop.value.indexOf(';') !== -1

      newFields[propName] = prop

      if (hasMultipleValues) {
        newFields[propName] = {
          ...prop,
          value: null,
          valueOptions: prop.value.split(';'),
        }
      }

      return newFields
    },
    {}
  )
}

export function maskFields(addressFields, rules) {
  return reduce(
    addressFields,
    (newAddressFields, prop, propName) => {
      const fieldRule = getField(propName, rules)

      newAddressFields[propName] = prop

      if (fieldRule && fieldRule.mask) {
        newAddressFields[propName] = {
          ...prop,
          ...(prop.value ? { value: msk(prop.value, fieldRule.mask) } : {}),
        }
      }

      return newAddressFields
    },
    {}
  )
}

export function addFocusToNextInvalidField(fields, rules) {
  const invalidFilledField = getFirstInvalidFilledField(fields, rules)

  if (invalidFilledField) {
    const { fieldName, field } = invalidFilledField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  const requiredField = getFirstRequiredFieldNotFilled(fields, rules)

  if (requiredField) {
    const { fieldName, field } = requiredField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  return fields
}

function getFirstInvalidFilledField(fields, rules) {
  const allFieldsVisited = addNewField(fields, 'visited', true)
  const validatedFields = validateChangedFields(
    allFieldsVisited,
    allFieldsVisited,
    rules
  )

  const orderedValidatedFields = orderFieldsByRules(validatedFields, rules)

  const firstInvalidField = find(
    orderedValidatedFields,
    field => field && field.valid === false
  )

  if (firstInvalidField) {
    return {
      fieldName: firstInvalidField.name,
      field: {
        ...validatedFields[firstInvalidField.name],
        focus: true,
      },
    }
  }

  return null
}

function orderFieldsByRules(fields, rules) {
  return reduce(
    fields,
    (acc, field, fieldName) => {
      const index = findIndex(
        rules.fields,
        ruleField => ruleField.name === fieldName
      )
      if (index === -1) return acc

      acc[index] = { ...field, name: fieldName }
      return acc
    },
    Array(Object.keys(fields).length)
  )
}

function getFirstRequiredFieldNotFilled(fields, rules) {
  const requiredFields = filter(rules.fields, field => field.required)
  const requiredFieldsNames = map(requiredFields, field => field.name)

  const fieldsNames = Object.keys(fields)
  const requiredFieldNotFilled = difference(requiredFieldsNames, fieldsNames)

  if (requiredFieldNotFilled && requiredFieldNotFilled.length > 0) {
    const nextRequiredFieldName = requiredFieldNotFilled[0]

    return { fieldName: nextRequiredFieldName, field: { focus: true } }
  }

  return null
}
