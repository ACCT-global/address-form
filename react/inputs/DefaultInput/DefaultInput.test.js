import React from 'react'
import renderer from 'react-test-renderer'

import address from '../../__mocks__/geolocationAddressValid'
import newAddress from '../../__mocks__/newAddress'

import Input from './index'

const GEOLOCATION_PROPS = {
  field: {
    name: 'number',
    size: 'small',
    maxLength: '9',
  },
  address: {
    ...address,
    number: {
      ...address.number,
      valid: true,
      loading: false,
    },
    postalCode: {
      value: '',
      loading: false,
      valid: true,
    },
  },
  onChange: jest.fn(),
  onBlur: jest.fn(),
}

const DEFAULT_PROPS = {
  field: {
    name: 'number',
    size: 'small',
    maxLength: '9',
  },
  addressId: {value: '1'},
  address: {
    addressId: {value: '1'},
    ...newAddress,
    number: {
      value: '',
      loading: false,
      valid: true,
    },
  },
  onChange: jest.fn(),
  onBlur: jest.fn(),
}

describe('Input', () => {
  it('should render default case', () => {
    const tree = renderer.create(<Input {...DEFAULT_PROPS} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render disabled input', () => {
    const tree = renderer.create(<Input {...GEOLOCATION_PROPS} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render custom class', () => {
    const props = {
      ...DEFAULT_PROPS,
      className: 'my-class',
    }
    const tree = renderer.create(<Input {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render different input type', () => {
    const props = {
      ...DEFAULT_PROPS,
      type: 'tel',
    }
    const tree = renderer.create(<Input {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render different id', () => {
    const props = {
      ...DEFAULT_PROPS,
      id: 'summary-postal-code',
    }
    const tree = renderer.create(<Input {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render different id based on fieldName', () => {
    const props = {
      ...DEFAULT_PROPS,
      id: 'my-context-{{fieldName}}',
    }
    const tree = renderer.create(<Input {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
