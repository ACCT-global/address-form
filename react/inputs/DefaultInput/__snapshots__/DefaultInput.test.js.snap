// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render custom class 1`] = `
<p
  className="input ship-number text"
>
  <label
    htmlFor="ship-number"
  />
  <input
    autoComplete="on"
    autoFocus={false}
    className="input-small success"
    data-hj-whitelist={true}
    disabled={false}
    id="ship-number"
    maxLength="9"
    name="number"
    onBlur={[MockFunction]}
    onChange={[Function]}
    onFocus={undefined}
    placeholder={undefined}
    type="text"
    value=""
  />
</p>
`;

exports[`Input should render default case 1`] = `
<p
  className="input ship-number text"
>
  <label
    htmlFor="ship-number"
  />
  <input
    autoComplete="on"
    autoFocus={false}
    className="input-small success"
    data-hj-whitelist={true}
    disabled={false}
    id="ship-number"
    maxLength="9"
    name="number"
    onBlur={[MockFunction]}
    onChange={[Function]}
    onFocus={undefined}
    placeholder={undefined}
    type="text"
    value=""
  />
</p>
`;

exports[`Input should render different id 1`] = `
<p
  className="input ship-number text"
>
  <label
    htmlFor="ship-number"
  />
  <input
    autoComplete="on"
    autoFocus={false}
    className="input-small success"
    data-hj-whitelist={true}
    disabled={false}
    id="ship-number"
    maxLength="9"
    name="number"
    onBlur={[MockFunction]}
    onChange={[Function]}
    onFocus={undefined}
    placeholder={undefined}
    type="text"
    value=""
  />
</p>
`;

exports[`Input should render different id based on fieldName 1`] = `
<p
  className="input ship-number text"
>
  <label
    htmlFor="ship-number"
  />
  <input
    autoComplete="on"
    autoFocus={false}
    className="input-small success"
    data-hj-whitelist={true}
    disabled={false}
    id="ship-number"
    maxLength="9"
    name="number"
    onBlur={[MockFunction]}
    onChange={[Function]}
    onFocus={undefined}
    placeholder={undefined}
    type="text"
    value=""
  />
</p>
`;

exports[`Input should render different input type 1`] = `
<p
  className="input ship-number text"
>
  <label
    htmlFor="ship-number"
  />
  <input
    autoComplete="on"
    autoFocus={false}
    className="input-small success"
    data-hj-whitelist={true}
    disabled={false}
    id="ship-number"
    maxLength="9"
    name="number"
    onBlur={[MockFunction]}
    onChange={[Function]}
    onFocus={undefined}
    placeholder={undefined}
    type="text"
    value=""
  />
</p>
`;

exports[`Input should render disabled input 1`] = `
<div
  className="input ship-notApplicable text"
  htmlFor="ship-number"
>
  <p
    className="input ship-number text"
  >
    <label
      htmlFor="ship-number"
    />
    <input
      autoComplete="on"
      autoFocus={false}
      className="input-small success"
      data-hj-whitelist={true}
      disabled={false}
      id="ship-number"
      maxLength="9"
      name="number"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onFocus={undefined}
      placeholder={undefined}
      type="text"
      value=""
    />
  </p>
  <p
    className="input ship-checkboxNumberLabel text"
  >
    <label
      htmlFor="ship-checkboxNumberLabel"
    />
    <input
      className="input ship-checkboxNumberLabel type"
      defaultChecked={undefined}
      id="ship-number"
      name="number"
      onBlur={[MockFunction]}
      onChange={[Function]}
      onFocus={undefined}
      placeholder={undefined}
      type="checkbox"
    />
  </p>
</div>
`;
